internal/infoschema_test.go

```diff
@@ -153,4 +139,237 @@ func TestProcessInfoSchema(t *testing.T) {
+			cols:  []string{"table_schema", "table_name"},
+			rows:  [][]driver.Value{{"public", "test"}},
+		}, {
+			query: "SELECT (.+) FROM information_schema.COLUMNS (.+)",
+			args:  []driver.Value{"public", "test"},
+			cols:  []string{"column_name", "data_type", "data_type", "is_nullable", "column_default", "character_maximum_length", "numeric_precision", "numeric_scale"},
+			rows: [][]driver.Value{
+				{"a", "text", nil, "NO", nil, nil, nil, nil},
+				{"b", "double precision", nil, "YES", nil, nil, 53, nil},
+				{"c", "bigint", nil, "YES", nil, nil, 64, 0}},
+		},
+		{
+			query: "SELECT (.+) FROM INFORMATION_SCHEMA.TABLE_CONSTRAINTS (.+)",
+			args:  []driver.Value{"public", "test"},
+			cols:  []string{"column_name", "constraint_type"},
+			rows:  [][]driver.Value{}, // No primary key --> force generation of synthetic key.
+		},
+		// Note: go-sqlmock mocks specify an ordered sequence
+		// of queries and results.  This (repeated) entry is
+		// needed because ProcessSqlData (redundantly) gets
+		// the set of tables via a SQL query.
+		{
+			query: "SELECT table_schema, table_name FROM information_schema.tables where table_type = 'BASE TABLE'",
+			cols:  []string{"table_schema", "table_name"},
+			rows:  [][]driver.Value{{"public", "test"}},
+		}, {
+			query: `SELECT [*] FROM "public"."test"`, // query is a regexp!
+			cols:  []string{"a", "b", "c"},
+			rows: [][]driver.Value{
+				{"cat", 42.3, nil},
+				{"dog", nil, 22}},
+		},
+	}
+	db := mkMockDB(t, ms)
+	conv := MakeConv()
+	err := ProcessInfoSchema(conv, db)
+	assert.Nil(t, err)
+	conv.SetDataMode()
+	var rows []spannerData
+	conv.SetDataSink(
+		func(table string, cols []string, vals []interface{}) {
+			rows = append(rows, spannerData{table: table, cols: cols, vals: vals})
+		})
+	ProcessSqlData(conv, db)
+	assert.Equal(t, []spannerData{
+		{table: "test", cols: []string{"a", "b", "synth_id"}, vals: []interface{}{"cat", float64(42.3), int64(0)}},
+		{table: "test", cols: []string{"a", "c", "synth_id"}, vals: []interface{}{"dog", int64(22), int64(-9223372036854775808)}}},
+		rows)
+	assert.Equal(t, int64(0), conv.Unexpecteds())
+}
+
+func TestSetRowStats(t *testing.T) {
+	ms := []mockSpec{
+		{
+			query: "SELECT table_schema, table_name FROM information_schema.tables where table_type = 'BASE TABLE'",
+			cols:  []string{"table_schema", "table_name"},
+			rows:  [][]driver.Value{{"public", "test1"}, {"public", "test2"}},
+		}, {
+			query: `SELECT COUNT[(][*][)] FROM "public"."test1"`,
+			cols:  []string{"count"},
+			rows:  [][]driver.Value{{5}},
+		}, {
+			query: `SELECT COUNT[(][*][)] FROM "public"."test2"`,
+			cols:  []string{"count"},
+			rows:  [][]driver.Value{{142}},
+		},
+	}
+	db := mkMockDB(t, ms)
+	conv := MakeConv()
+	conv.SetDataMode()
+	SetRowStats(conv, db)
+	assert.Equal(t, int64(5), conv.stats.rows["test1"])
+	assert.Equal(t, int64(142), conv.stats.rows["test2"])
+	assert.Equal(t, int64(0), conv.Unexpecteds())
+}
+
+type mockSpec struct {
```

Comment:
Maybe consider to put this on the top for a better reading flow.

Reply:
Thanks - done!